internal/cmd/gocdk/internal/static/static.go

```diff
@@ -14,12 +14,285 @@
+	Path string
+	// IsTemplate is true means that the source file will be parsed and executed
+	// as a Go template (see https://golang.org/pkg/text/template/). The template
+	// will be passed templateData. The result of executing the template will be
+	// used as the content instead of the raw source.
+	// NOTE: We can't just treat every file as a template; some content files
+	// include template directives that should not be interpolated at copy time.
+	// For example, the .go files for demos include constants for the HTML they
+	// generate which include template directives. We chose to default to treating
+	// source files as raw bytes rather than as template because most files are
+	// copied as is, and it would be surprising if including a template directive
+	// accidentally (e.g., a stray "{{") caused things to break.
+	// TODO(rvangent): Consider specifying a list of template keys that the
+	// template requires, and enforcing that they are provided in TemplateData.
+	// TemplateData would probably have to always be a map to make that possible.
+	// Alternatively, consider at least verifying that if TemplateData was
+	// provided, it was used, and if IsTemplate=true, that TemplateData is
+	// not nil.
+	IsTemplate bool
+
+	// AddTo sets a destination file that the content from this file will be
+	// added to. The destination file must already exist. See AddToAfterMarker
+	// for where the content will be added, and AddToSkipMarker to identify if
+	// the content has already been added.
+	AddTo string
+
+	// AddToSkipMarker is used to determine whether the addition to the
+	// destination file configured by AddTo has already occurred, and thus this
+	// source file should be skipped. It defaults to the content.
+	AddToSkipMarker string
+
+	// AddToMarker determines where in the destinatation file configured by AddTo
+	// the content is placed. If AddToMarker is the empty string, the content is
+	// appended. Otherwise, it is inserted into the file at the next line after
+	// the marker. If the marker is not found, an error is returned.
+	AddToAfterMarker string
+
+	// AddToDescription describes the content that was added; used when logging
+	// changes via MaterializeOptions.Logger. Log messages are of the form
+	// "added <AddToDescription> to <filename>". Defaults to "content".
+	AddToDescription string
+}
+
+// Materialize recursively materializes files from the static data at srcRoot to
+// dstRoot.
+//
+// By default, it copies files. If a file already exists at the destination,
+// Materialize will return an error unless force is true.
+//
+// Additional behavior can be configured by adding a JSON file to the static
+// data, at srcRoot + ".json". The JSON file should consist of a sequence of
+// JSON objects that will be deserialized into fileOptions structs; see
+// the struct docstring for more info.
+func Materialize(dstRoot, srcRoot string, opts *MaterializeOptions) error {
```

Comment:
An alternative to the `.json` files with configuration would be to do that in Go, and pass it here. So, each caller would construct a map of path -> file options and pass it to `Materialize`. I'm starting to lean more toward that approach because it will probably work better for things in the TODOs above like shared resources and prompts. Feedback welcome.